src/TraitCollection.php

```
@@ -0,0 +1,159 @@
+<?php
+
+namespace Raml;
+
+use Exception;
+
+/**
+ *  Singleton class used to register all traits in one place
+ **/
+class TraitCollection implements \Iterator
+{
+    /**
+     * Hold the class instance.
+     *
+     * @var self
+     **/
+    private static $instance = null;
+
+    /**
+     * Collection
+     *
+     * @var array
+     **/
+    private $collection = [];
+
+    /**
+     * Current position
+     *
+     * @var string
+     **/
+    private $position = 0;
+
+    /**
+    * prevent initiation from outside, there can be only one!
+    *
+    **/
+    private function __construct()
+    {
+        $this->collection = [];
+        $this->position = 0;
+    }
+ 
+    /**
+     * The object is created from within the class itself
+     * only if the class has no instance.
+     *
+     * @return TraitCollection
+     **/
+    public static function getInstance()
+    {
+        if (self::$instance == null) {
+            self::$instance = new TraitCollection();
+        }
+    
+        return self::$instance;
+    }
+
+    public function current()
+    {
+        return $this->collection[$this->position];
```

This needs defensive check that the position actually exists in the collection. You can reuse the `valid` method here and throw an appropriate exception when `position` is not set.